File: explorer/config/operation_config.py
from explorer.services.explorer_cache_service import ExplorerCacheService
from shared.services.universal_raw_service import UniversalRawService
from shared.services.universal_enriched_service import UniversalEnrichedService
from shared.services.universal_metric_service import UniversalMetricService
from shared.services.universal_list_service import UniversalListService
from shared.services.snapshots_service import SnapshotsService
from explorer.util import operation_util

OPERATION_DEFINITIONS = {
    # Universal Operations
    "init_user": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "raw",
        "handler": lambda user_id, **kwargs: UniversalRawService().get_user_universal(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: None,
        "setup": lambda user_id, op: ExplorerCacheService().create_empty_operation_chain_cache(user_id=user_id)
    },
    "reset": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "raw",
        "handler": lambda user_id, **kwargs: UniversalRawService().get_user_universal(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: None,
        "setup": lambda user_id, op: ExplorerCacheService().create_empty_operation_chain_cache(user_id=user_id)
    },
    "filter": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name", "filter_value", "filter_type"],
        "operation_expected_result_data_type": "raw",
        "handler": lambda user_id, data_source, column_name, filter_value, filter_type, **kwargs: UniversalRawService().filter(
            user_id=user_id, data_source=data_source, column_name=column_name, filter_value=filter_value, filter_type=filter_type, **kwargs
        ),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "traverse": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "start_id", "traversal_directions"],
        "operation_expected_result_data_type": "raw",
        "handler": lambda user_id, data_source, start_id, traversal_directions, **kwargs: UniversalRawService().traverse(
            user_id=user_id, data_source=data_source, start_id=start_id, traversal_directions=traversal_directions, **kwargs
        ),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "undo": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "raw",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": lambda user_id, op: ExplorerCacheService().delete_most_recent_operation_from_chain(user_id=user_id),
        "precondition": {
                "condition": lambda user_id: len(ExplorerCacheService().get_operation_chain(user_id=user_id)) > 1,
                "error_message": "Cannot undo operation: no operations to undo"
            }
    },
    "load_snapshot": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "snapshot_id"],
        "operation_expected_result_data_type": "raw",
        # Here we wrap the utility method that assembles the dataset list.
        "handler": lambda user_id, **kwargs: operation_util.assemble_dataset_list_from_operation_chain(user_id, operation_chain=kwargs.get("data_source")),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: SnapshotsService().get_snapshot_operation_chain(
            user_id=user_id,
            snapshot_id=op.operation_arguments["snapshot_id"]
        ),
        "setup": lambda user_id, op: ExplorerCacheService().empty_operation_chain(user_id=user_id)
    },
    "group_aggregate": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "group_columns", "aggregate_operation", "target_column"],
        "operation_expected_result_data_type": "enriched",
        "handler": lambda user_id, **kwargs: UniversalEnrichedService().group_aggregate(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_unique_column_values": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name", "options_call"],
        "operation_expected_result_data_type": "list",
        "handler": lambda user_id, **kwargs: UniversalListService().get_unique_column_values(user_id=user_id, **kwargs),
        "cache_policy": lambda op: not op.operation_arguments.get("options_call", False),
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_unique_json_keys": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "options_call"],
        "operation_expected_result_data_type": "list",
        "handler": lambda user_id, **kwargs: UniversalListService().get_unique_json_keys(user_id=user_id, **kwargs),
        "cache_policy": lambda op: not op.operation_arguments.get("options_call", False),
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_unique_json_values": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "options_call"],
        "operation_expected_result_data_type": "list",
        "handler": lambda user_id, **kwargs: UniversalListService().get_unique_json_values(user_id=user_id, **kwargs),
        "cache_policy": lambda op: not op.operation_arguments.get("options_call", False),
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_unique_json_key_values": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "json_key", "options_call"],
        "operation_expected_result_data_type": "list",
        "handler": lambda user_id, **kwargs: UniversalListService().get_unique_json_key_values(user_id=user_id, **kwargs),
        "cache_policy": lambda op: not op.operation_arguments.get("options_call", False),
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_count": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name"],
        "operation_expected_result_data_type": "metric",
        "handler": lambda user_id, **kwargs: UniversalMetricService().get_count(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_average": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name"],
        "operation_expected_result_data_type": "metric",
        "handler": lambda user_id, **kwargs: UniversalMetricService().get_average(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_sum": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name"],
        "operation_expected_result_data_type": "metric",
        "handler": lambda user_id, **kwargs: UniversalMetricService().get_sum(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_min": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name"],
        "operation_expected_result_data_type": "metric",
        "handler": lambda user_id, **kwargs: UniversalMetricService().get_min(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },
    "get_max": {
        "operation_type": "universal",
        "operation_expected_arguments": ["user_id", "column_name"],
        "operation_expected_result_data_type": "metric",
        "handler": lambda user_id, **kwargs: UniversalMetricService().get_max(user_id=user_id, **kwargs),
        "cache_policy": lambda op: True,
        "data_source": lambda user_id, op: ExplorerCacheService().get_most_recent_operation_chain_raw_data_result(user_id=user_id),
        "setup": None
    },

    # State Operations
    "start_explorer_session": {
        "operation_type": "state",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "status",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().create_empty_operation_chain_cache(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "end_explorer_session": {
        "operation_type": "state",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "status",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().delete_operation_chain_cache(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "get_operation_chain": {
        "operation_type": "state",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "operation_chain",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().get_operation_chain(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "get_operation_chain_operations": {
        "operation_type": "state",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "operations_list",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().extract_operation_chain_operations(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "get_operation_chain_results": {
        "operation_type": "state",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "results_list",
        "handler": lambda user_id, **kwargs: ExplorerCacheService().extract_operation_chain_result_data(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },

    # Snapshot Operations
    "create_snapshot": {
        "operation_type": "snapshot",
        "operation_expected_arguments": ["user_id", "title"],
        "operation_expected_result_data_type": "snapshot",
        "handler": lambda user_id, data_source, **kwargs: SnapshotsService().create_snapshot(user_id=user_id, operation_chain=data_source, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: ExplorerCacheService().extract_operation_chain_operations(user_id=user_id),
        "setup": None
    },
    "delete_snapshot": {
        "operation_type": "snapshot",
        "operation_expected_arguments": ["user_id", "snapshot_id"],
        "operation_expected_result_data_type": "status",
        "handler": lambda user_id, **kwargs: SnapshotsService().delete_snapshot(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "update_snapshot": {
        "operation_type": "snapshot",
        "operation_expected_arguments": ["user_id", "snapshot_id"],
        "operation_expected_result_data_type": "snapshot",
        "handler": lambda user_id, **kwargs: SnapshotsService().update_snapshot(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "get_snapshot": {
        "operation_type": "snapshot",
        "operation_expected_arguments": ["user_id", "snapshot_id"],
        "operation_expected_result_data_type": "snapshot",
        "handler": lambda user_id, **kwargs: SnapshotsService().get_snapshot(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    },
    "get_all_snapshots": {
        "operation_type": "snapshot",
        "operation_expected_arguments": ["user_id"],
        "operation_expected_result_data_type": "snapshot_list",
        "handler": lambda user_id, **kwargs: SnapshotsService().get_all_snapshots(user_id=user_id, **kwargs),
        "cache_policy": lambda op: False,
        "data_source": lambda user_id, op: None,
        "setup": None
    }
}

File: frontend/src/explorer/config/OperationConfig.js
export const OPERATION_CONFIG = {
    // Universal operations
    get_full_data: {
      op_name:    "get_full_data",
      op_args:    [],
      httpMethod: "GET",
    },
    filter: {
      op_name:    "filter",
      op_args:    ["column_name", "filter_value", "filter_type"],
      httpMethod: "GET",
    },
    traverse: {
      op_name:    "traverse",
      op_args:    ["start_id", "traversal_directions"],
      httpMethod: "GET",
    },
    group_aggregate: {
      op_name:    "group_aggregate",
      op_args:    ["group_columns", "aggregate_operation", "target_column", "frequency"],
      httpMethod: "GET",
    },
    get_unique_column_values: {
      op_name:    "get_unique_column_values",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_unique_json_keys: {
      op_name:    "get_unique_json_keys",
      op_args:    [],
      httpMethod: "GET",
    },
    get_unique_json_values: {
      op_name:    "get_unique_json_values",
      op_args:    [],
      httpMethod: "GET",
    },
    get_unique_json_key_values: {
      op_name:    "get_unique_json_key_values",
      op_args:    ["json_key"],
      httpMethod: "GET",
    },
  
    get_unique_column_values_filter_options: {
      op_name:    "get_unique_column_values_filter_options",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_unique_json_keys_filter_options: {
      op_name:    "get_unique_json_keys_filter_options",
      op_args:    [],
      httpMethod: "GET",
    },
    get_unique_json_values_filter_options: {
      op_name:    "get_unique_json_values_filter_options",
      op_args:    [],
      httpMethod: "GET",
    },
    get_unique_json_key_values_filter_options: {
      op_name:    "get_unique_json_key_values_filter_options",
      op_args:    ["json_key"],
      httpMethod: "GET",
    },
  
    get_count: {
      op_name:    "get_count",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_average: {
      op_name:    "get_average",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_sum: {
      op_name:    "get_sum",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_min: {
      op_name:    "get_min",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
    get_max: {
      op_name:    "get_max",
      op_args:    ["column_name"],
      httpMethod: "GET",
    },
  
    // State operations
    start_explorer_session: {
      op_name:    "start_explorer_session",
      op_args:    [],
      httpMethod: "GET",
    },
    reset_explorer_session: {
      op_name:    "reset_explorer_session",
      op_args:    [],
      httpMethod: "GET",
    },
    undo_operation: {
      op_name:    "undo_operation",
      op_args:    [],
      httpMethod: "POST",
    },
    load_snapshot: {
      op_name:    "load_snapshot",
      op_args:    ["snapshot_id"],
      httpMethod: "GET",
    },
    end_explorer_session: {
      op_name:    "end_explorer_session",
      op_args:    [],
      httpMethod: "POST",
    },
  
    // Snapshot operations
    save_snapshot: {
      op_name:    "save_snapshot",
      op_args:    ["title", "description"],
      httpMethod: "POST",
    },
    delete_snapshot: {
      op_name:    "delete_snapshot",
      op_args:    ["snapshot_id"],
      httpMethod: "DELETE",
    },
    update_snapshot: {
      op_name:    "update_snapshot",
      op_args:    ["snapshot_id", "title", "description"],
      httpMethod: "PUT",
    },
    get_snapshot: {
      op_name:    "get_snapshot",
      op_args:    ["snapshot_id"],
      httpMethod: "GET",
    },
    get_all_snapshots: {
      op_name:    "get_all_snapshots",
      op_args:    [],
      httpMethod: "GET",
    },
  
    // Utility / operations chain
    get_operation_chain: {
      op_name:    "get_operation_chain",
      op_args:    [],
      httpMethod: "GET",
    },
    get_operation_chain_operations: {
      op_name:    "get_operation_chain_operations",
      op_args:    [],
      httpMethod: "GET",
    },
    get_operation_chain_results: {
      op_name:    "get_operation_chain_results",
      op_args:    [],
      httpMethod: "GET",
    },
    get_column_operator_options: {
      op_name:    "get_column_operator_options",
      op_args:    [],
      httpMethod: "GET",
    },
    get_filterable_column_names: {
      op_name:    "get_filterable_column_names",
      op_args:    [],
      httpMethod: "GET",
    },
    get_operation_types: {
      op_name:    "get_operation_types",
      op_args:    [],
      httpMethod: "GET",
    },
    get_operation_names_for_result_data_type: {
      op_name:    "get_operation_names_for_result_data_type",
      op_args:    ["operation_result_data_type"],
      httpMethod: "GET",
    },
    get_operation_argument_names: {
      op_name:    "get_operation_argument_names",
      op_args:    ["operation_name"],
      httpMethod: "GET",
    },
    get_operation_argument_options: {
      op_name:    "get_operation_argument_options",
      op_args:    ["operation_name", "operation_argument_name", "prev_argument_values"],
      httpMethod: "GET",
    },
  };